client/src/main/scala/org/apache/celeborn/client/RequestLocationCallContext.scala

```diff
@@ -17,23 +17,56 @@
 
 package org.apache.celeborn.client
 
+import java.util
+import java.util.concurrent.ConcurrentHashMap
+
+import org.apache.celeborn.common.internal.Logging
 import org.apache.celeborn.common.protocol.PartitionLocation
 import org.apache.celeborn.common.protocol.message.ControlMessages.{ChangeLocationResponse, RegisterShuffleResponse}
 import org.apache.celeborn.common.protocol.message.StatusCode
 import org.apache.celeborn.common.rpc.RpcCallContext
 
 trait RequestLocationCallContext {
-  def reply(status: StatusCode, partitionLocationOpt: Option[PartitionLocation]): Unit
+  def reply(
+      partitionId: Int,
+      status: StatusCode,
+      partitionLocationOpt: Option[PartitionLocation]): Unit
 }
 
-case class ChangeLocationCallContext(context: RpcCallContext) extends RequestLocationCallContext {
-  override def reply(status: StatusCode, partitionLocationOpt: Option[PartitionLocation]): Unit = {
-    context.reply(ChangeLocationResponse(status, partitionLocationOpt))
+case class ChangeLocationsCallContext(
+    context: RpcCallContext,
+    partitionCount: Int)
+  extends RequestLocationCallContext with Logging {
+  val endedMapIds = new util.HashSet[Integer]()
+  val newLocs = new ConcurrentHashMap[Integer, (StatusCode, PartitionLocation)](partitionCount)
+
+  def markMapperEnd(mapId: Int): Unit = this.synchronized {
+    endedMapIds.add(mapId)
+  }
+
+  override def reply(
+      partitionId: Int,
+      status: StatusCode,
+      partitionLocationOpt: Option[PartitionLocation]): Unit = this.synchronized {
+    newLocs.putIfAbsent(
+      partitionId,
+      (status, partitionLocationOpt.getOrElse(new PartitionLocation())))
```

Comment:
Better not transfer `unknown partitionlocation` to client, as is useless.

Reply:
done